<span style= "font-size:46px; color:yellowgreen"> **Sprint 10:** 🐍 ---Introducció a 🐼s!---- 🏃🏿‍♂️‍➡️ 


<span style="font-size:16px; color: chartreuse">*Author*: **Dioulde Diao** </span> <br>
<span style="font-size:14px; color: green">*Peer reviewed by*: **Francesco Eirado Enes** </span>

<span style="font-size:40px; color: paleturquoise">***Nivell 1*** </span>

<span style="font-size:16px; color:maroon"> **1. Importa com un DataFrame l'arxiu sprint10.xlsx. Assegura't que el fitxer s'importa correctament, amb els noms de columnes que li corresponen, sense manipular l'arxiu original.** </span>

* <span style="font-size:14px">Ordena el DataFrame pel país d'origen. En cas d'empat, ordena pel nom de la ciutat.
* <span style="font-size:14px">Mostra les primeres 10 files.
* <span style="font-size:14px">Addicionalment, fes un print on comprovi que el DNI només té valors únics.

In [132]:
import pandas as pd
treballadors_2025=pd.read_excel("sprint10.xlsx", index_col=0, header=3, na_values="Missing value", keep_default_na=False)
treballadors_2025

,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional
0,Inês,Ferreira Silva,16928694K,Portugal,Lisboa,25,2,1953,D,1.144 €,,True,Grup B
1,Clara,Sánchez Martínez,27724652S,Espanya,Barcelona,18,3,1996,D,1.253 €,True,,Grup A
2,Fatima,Fassi,38141675A,Marroc,Rabat,6,11,2005,A,1.441 €,True,,Grup A
3,Khadija,Bennani Bennani,59157262R,Marroc,Rabat,20,1,1995,D,1.944 €,,True,Grup B
4,Toni,Sánchez García,69630528M,Espanya,Barcelona,9,8,1999,H,1.043 €,,True,Grup A
...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,Marta,Ferrer Ferrer,25161375F,Espanya,Sevilla,1,6,1951,D,1.216 €,,True,Grup B
996,Joan,García,52145541P,Espanya,Sevilla,11,4,1959,H,971 €,,True,Grup A
997,Laia,Ferrer Martínez,69760120X,Espanya,Barcelona,11,11,1980,D,682 €,,True,Grup A
998,Jordi,García,82947791W,Espanya,Barcelona,23,5,1984,H,1.699 €,True,,Grup C


In [133]:
treballadors_2025.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, 0 to 999
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Nom                1000 non-null   object
 1   Cognoms            1000 non-null   object
 2   DNI                1000 non-null   object
 3   País d'origen      1000 non-null   object
 4   Ciutat             1000 non-null   object
 5   Dia de Naixement   1000 non-null   int64 
 6   Mes de Naixement   1000 non-null   int64 
 7   Any de Naixement   1000 non-null   int64 
 8   Gènere             1000 non-null   object
 9   Salari mensual     1000 non-null   object
 10  Fills              1000 non-null   object
 11  No Fills           1000 non-null   object
 12  Grup Professional  1000 non-null   object
dtypes: int64(3), object(10)
memory usage: 109.4+ KB


In [134]:
print(f"Estamos trabajando con la versión de pandas {pd.__version__}")

Estamos trabajando con la versión de pandas 2.3.3


* <span style="font-size:14px">Ordena el DataFrame pel país d'origen. En cas d'empat, ordena pel nom de la ciutat.

In [135]:
treballadors_2025=treballadors_2025.sort_values(by=["País d'origen", "Ciutat"], ascending=True)

* <span style="font-size:14px">Mostra les primeres 10 files.

In [136]:
treballadors_2025.head(10)

,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional
21,Mia,Schneider Fischer,28973553Z,Alemanya,Berlín,22,10,1976,A,951 €,,True,Grup A
154,Laura,Schneider Fischer,37399141L,Alemanya,Berlín,2,2,1958,D,1.769 €,True,,Grup B
224,Lea,Schneider Schneider,37368317L,Alemanya,Berlín,23,10,2005,D,2.013 €,,True,Grup B
278,Mia,Fischer,21390098Z,Alemanya,Berlín,11,8,1950,D,1.557 €,True,,Grup B
602,Jonas,Schneider,44060014R,Alemanya,Berlín,22,11,1985,H,2.754 €,True,,Grup D
871,Lea,Fischer,14773153R,Alemanya,Berlín,9,9,1986,D,1.370 €,True,,Grup A
281,Lea,Müller,23266650S,Alemanya,Hamburg,14,4,2003,D,1.314 €,,True,Grup A
435,Anna,Müller,83274277X,Alemanya,Hamburg,1,1,1987,D,2.464 €,,True,Grup C
444,Laura,Schmidt Müller,60161784X,Alemanya,Hamburg,15,6,1987,NC,2.035 €,True,,Grup C
487,Lukas,Müller Fischer,60982309S,Alemanya,Hamburg,28,3,1971,H,2.042 €,,True,Grup B


* <span style="font-size:14px">Addicionalment, fes un print on comprovi que el DNI només té valors únics.

In [137]:
#Forma més completa de comprovar les característiques d'un df. ** És la que més m'agrada**
treballadors_2025["DNI"].describe()

#Alternatives a describe: (Eliminar # de davant la query per executar)

#1) 
#treballadors_2025["DNI"].nunique()
    # Podem printar els valors únics fent:
#treballadors_2025["DNI"].unique() #Retorna un array amb tots els valors únics.

#2)
#treballadors_2025["DNI"].value_counts() #Retorna un df amb el recompta de les files amb valors únics. En l'informació de la pàgina podem veure què hi ha 1000 files i que totes tenen valors únics.  


count          1000
unique         1000
top       28973553Z
freq              1
Name: DNI, dtype: object

<span style="font-size:16px; color: maroon">**Ex.2.** </span>

* <span style="font-size:14px">Crea una columna que sigui el nom complet.
* <span style="font-size:14px">Crea una columna si la persona és nascuda a Espanya o no.
* <span style="font-size:14px">Posa el DNI com a índex del DataFrame (noms de files).
* <span style="font-size:14px">Substitueix el nom de les columnes Dia de Naixement, Mes de Naixement i Any de Naixement per Dia, Mes i Any.
* <span style="font-size:14px">Substitueix H per Home, D per Dona, A per Altres i NC per una dada faltant (nan/null/na).

<span style="font-size:16px; color: salmon">**Mostra tots els canvis que has realitzat en una sola taula.**

* <span style="font-size:14px">Crea una columna que sigui el nom complet.

In [138]:
treballadors_2025["Nom Cognoms"]=treballadors_2025["Nom"]+" "+treballadors_2025["Cognoms"]
treballadors_2025

,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Nom Cognoms
21,Mia,Schneider Fischer,28973553Z,Alemanya,Berlín,22,10,1976,A,951 €,,True,Grup A,Mia Schneider Fischer
154,Laura,Schneider Fischer,37399141L,Alemanya,Berlín,2,2,1958,D,1.769 €,True,,Grup B,Laura Schneider Fischer
224,Lea,Schneider Schneider,37368317L,Alemanya,Berlín,23,10,2005,D,2.013 €,,True,Grup B,Lea Schneider Schneider
278,Mia,Fischer,21390098Z,Alemanya,Berlín,11,8,1950,D,1.557 €,True,,Grup B,Mia Fischer
602,Jonas,Schneider,44060014R,Alemanya,Berlín,22,11,1985,H,2.754 €,True,,Grup D,Jonas Schneider
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
547,Emily,Taylor Jones,89577876S,Regne Unit,Manchester,28,3,1958,D,2.033 €,,True,Grup B,Emily Taylor Jones
728,George,Brown Jones,57441590Y,Regne Unit,Manchester,27,12,1979,H,1.130 €,True,,Grup A,George Brown Jones
751,Olivia,Brown Brown,58204038A,Regne Unit,Manchester,28,8,1952,A,1.023 €,,True,Grup A,Olivia Brown Brown
854,Isla,Jones Brown,28367234K,Regne Unit,Manchester,28,3,1999,D,1.197 €,,True,Grup A,Isla Jones Brown


* <span style="font-size:14px">Crea una columna si la persona és nascuda a Espanya o no.

In [139]:
treballadors_2025["Nascut_Espanya"]=treballadors_2025["País d'origen"]=="Espanya"
treballadors_2025

,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Nom Cognoms,Nascut_Espanya
21,Mia,Schneider Fischer,28973553Z,Alemanya,Berlín,22,10,1976,A,951 €,,True,Grup A,Mia Schneider Fischer,False
154,Laura,Schneider Fischer,37399141L,Alemanya,Berlín,2,2,1958,D,1.769 €,True,,Grup B,Laura Schneider Fischer,False
224,Lea,Schneider Schneider,37368317L,Alemanya,Berlín,23,10,2005,D,2.013 €,,True,Grup B,Lea Schneider Schneider,False
278,Mia,Fischer,21390098Z,Alemanya,Berlín,11,8,1950,D,1.557 €,True,,Grup B,Mia Fischer,False
602,Jonas,Schneider,44060014R,Alemanya,Berlín,22,11,1985,H,2.754 €,True,,Grup D,Jonas Schneider,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
547,Emily,Taylor Jones,89577876S,Regne Unit,Manchester,28,3,1958,D,2.033 €,,True,Grup B,Emily Taylor Jones,False
728,George,Brown Jones,57441590Y,Regne Unit,Manchester,27,12,1979,H,1.130 €,True,,Grup A,George Brown Jones,False
751,Olivia,Brown Brown,58204038A,Regne Unit,Manchester,28,8,1952,A,1.023 €,,True,Grup A,Olivia Brown Brown,False
854,Isla,Jones Brown,28367234K,Regne Unit,Manchester,28,3,1999,D,1.197 €,,True,Grup A,Isla Jones Brown,False


* <span style="font-size:14px">Posa el DNI com a índex del DataFrame (noms de files).

In [140]:
treb=treballadors_2025.set_index("DNI")
treb

,Nom,Cognoms,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Nom Cognoms,Nascut_Espanya
DNI,,,,,,,,,,,,,,
28973553Z,Mia,Schneider Fischer,Alemanya,Berlín,22,10,1976,A,951 €,,True,Grup A,Mia Schneider Fischer,False
37399141L,Laura,Schneider Fischer,Alemanya,Berlín,2,2,1958,D,1.769 €,True,,Grup B,Laura Schneider Fischer,False
37368317L,Lea,Schneider Schneider,Alemanya,Berlín,23,10,2005,D,2.013 €,,True,Grup B,Lea Schneider Schneider,False
21390098Z,Mia,Fischer,Alemanya,Berlín,11,8,1950,D,1.557 €,True,,Grup B,Mia Fischer,False
44060014R,Jonas,Schneider,Alemanya,Berlín,22,11,1985,H,2.754 €,True,,Grup D,Jonas Schneider,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
89577876S,Emily,Taylor Jones,Regne Unit,Manchester,28,3,1958,D,2.033 €,,True,Grup B,Emily Taylor Jones,False
57441590Y,George,Brown Jones,Regne Unit,Manchester,27,12,1979,H,1.130 €,True,,Grup A,George Brown Jones,False
58204038A,Olivia,Brown Brown,Regne Unit,Manchester,28,8,1952,A,1.023 €,,True,Grup A,Olivia Brown Brown,False


* <span style="font-size:14px">Substitueix el nom de les columnes Dia de Naixement, Mes de Naixement i Any de Naixement per Dia, Mes i Any.

In [141]:
treb=treb.rename(columns={"Dia de Naixement":"Dia", "Mes de Naixement":"Mes", "Any de Naixement":"Any"})
treb

,Nom,Cognoms,País d'origen,Ciutat,Dia,Mes,Any,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Nom Cognoms,Nascut_Espanya
DNI,,,,,,,,,,,,,,
28973553Z,Mia,Schneider Fischer,Alemanya,Berlín,22,10,1976,A,951 €,,True,Grup A,Mia Schneider Fischer,False
37399141L,Laura,Schneider Fischer,Alemanya,Berlín,2,2,1958,D,1.769 €,True,,Grup B,Laura Schneider Fischer,False
37368317L,Lea,Schneider Schneider,Alemanya,Berlín,23,10,2005,D,2.013 €,,True,Grup B,Lea Schneider Schneider,False
21390098Z,Mia,Fischer,Alemanya,Berlín,11,8,1950,D,1.557 €,True,,Grup B,Mia Fischer,False
44060014R,Jonas,Schneider,Alemanya,Berlín,22,11,1985,H,2.754 €,True,,Grup D,Jonas Schneider,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
89577876S,Emily,Taylor Jones,Regne Unit,Manchester,28,3,1958,D,2.033 €,,True,Grup B,Emily Taylor Jones,False
57441590Y,George,Brown Jones,Regne Unit,Manchester,27,12,1979,H,1.130 €,True,,Grup A,George Brown Jones,False
58204038A,Olivia,Brown Brown,Regne Unit,Manchester,28,8,1952,A,1.023 €,,True,Grup A,Olivia Brown Brown,False


* <span style="font-size:14px">Substitueix H per Home, D per Dona, A per Altres i NC per una dada faltant (nan/null/na).

In [142]:
treb["Gènere"]=treb["Gènere"].replace({"H":"Home", "D":"Dona", "A":"Altres", "NC":"NaN"})
treb

,Nom,Cognoms,País d'origen,Ciutat,Dia,Mes,Any,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Nom Cognoms,Nascut_Espanya
DNI,,,,,,,,,,,,,,
28973553Z,Mia,Schneider Fischer,Alemanya,Berlín,22,10,1976,Altres,951 €,,True,Grup A,Mia Schneider Fischer,False
37399141L,Laura,Schneider Fischer,Alemanya,Berlín,2,2,1958,Dona,1.769 €,True,,Grup B,Laura Schneider Fischer,False
37368317L,Lea,Schneider Schneider,Alemanya,Berlín,23,10,2005,Dona,2.013 €,,True,Grup B,Lea Schneider Schneider,False
21390098Z,Mia,Fischer,Alemanya,Berlín,11,8,1950,Dona,1.557 €,True,,Grup B,Mia Fischer,False
44060014R,Jonas,Schneider,Alemanya,Berlín,22,11,1985,Home,2.754 €,True,,Grup D,Jonas Schneider,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
89577876S,Emily,Taylor Jones,Regne Unit,Manchester,28,3,1958,Dona,2.033 €,,True,Grup B,Emily Taylor Jones,False
57441590Y,George,Brown Jones,Regne Unit,Manchester,27,12,1979,Home,1.130 €,True,,Grup A,George Brown Jones,False
58204038A,Olivia,Brown Brown,Regne Unit,Manchester,28,8,1952,Altres,1.023 €,,True,Grup A,Olivia Brown Brown,False


<span style="font-size:16px; color: salmon">**Mostra tots els canvis que has realitzat en una sola taula.**

In [143]:
treb=treb.drop(columns=["Nom", "Cognoms"])
treb

,País d'origen,Ciutat,Dia,Mes,Any,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Nom Cognoms,Nascut_Espanya
DNI,,,,,,,,,,,,
28973553Z,Alemanya,Berlín,22,10,1976,Altres,951 €,,True,Grup A,Mia Schneider Fischer,False
37399141L,Alemanya,Berlín,2,2,1958,Dona,1.769 €,True,,Grup B,Laura Schneider Fischer,False
37368317L,Alemanya,Berlín,23,10,2005,Dona,2.013 €,,True,Grup B,Lea Schneider Schneider,False
21390098Z,Alemanya,Berlín,11,8,1950,Dona,1.557 €,True,,Grup B,Mia Fischer,False
44060014R,Alemanya,Berlín,22,11,1985,Home,2.754 €,True,,Grup D,Jonas Schneider,False
...,...,...,...,...,...,...,...,...,...,...,...,...
89577876S,Regne Unit,Manchester,28,3,1958,Dona,2.033 €,,True,Grup B,Emily Taylor Jones,False
57441590Y,Regne Unit,Manchester,27,12,1979,Home,1.130 €,True,,Grup A,George Brown Jones,False
58204038A,Regne Unit,Manchester,28,8,1952,Altres,1.023 €,,True,Grup A,Olivia Brown Brown,False


<span style="font-size:16px; color:maroon">**Ex.3.**</span>

* <span style="font-size:14px">Junta les columnes Fills i No Fills en una sola columna, utilitzant el mètode <font color="azure"> *.apply()* </font> i definint una funció que resolgui el problema. La columna nova ha de dir-se "Fills" i prendre els valors "Sí" o "No". </span>

In [144]:
treb["Fills"]=treb[["No Fills"]].apply(
    lambda x: "No" if x.fillna(False).any() else "Sí", axis=1) #axis=1 indica que la función se aplica a cada fila (en lugar de a cada columna).
treb=treb.drop(columns=["No Fills"])
treb

C:\Users\dioul\AppData\Local\Temp\ipykernel_31788\3596441969.py:2: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  lambda x: "No" if x.fillna(False).any() else "Sí", axis=1) #axis=1 indica que la función se aplica a cada fila (en lugar de a cada columna).


,País d'origen,Ciutat,Dia,Mes,Any,Gènere,Salari mensual,Fills,Grup Professional,Nom Cognoms,Nascut_Espanya
DNI,,,,,,,,,,,
28973553Z,Alemanya,Berlín,22,10,1976,Altres,951 €,No,Grup A,Mia Schneider Fischer,False
37399141L,Alemanya,Berlín,2,2,1958,Dona,1.769 €,Sí,Grup B,Laura Schneider Fischer,False
37368317L,Alemanya,Berlín,23,10,2005,Dona,2.013 €,No,Grup B,Lea Schneider Schneider,False
21390098Z,Alemanya,Berlín,11,8,1950,Dona,1.557 €,Sí,Grup B,Mia Fischer,False
44060014R,Alemanya,Berlín,22,11,1985,Home,2.754 €,Sí,Grup D,Jonas Schneider,False
...,...,...,...,...,...,...,...,...,...,...,...
89577876S,Regne Unit,Manchester,28,3,1958,Dona,2.033 €,No,Grup B,Emily Taylor Jones,False
57441590Y,Regne Unit,Manchester,27,12,1979,Home,1.130 €,Sí,Grup A,George Brown Jones,False
58204038A,Regne Unit,Manchester,28,8,1952,Altres,1.023 €,No,Grup A,Olivia Brown Brown,False


Qué representa la <span style="color:red; font-size:12px">**lambda**</span> en la consulta/funció de dalt?<br>
* <span style="font-size:12px"> https://makeitrealcamp.gitbook.io/guias-de-make-it-real/pandas/la-funcion-lambda </span>

##### Reordenem el df: 

In [145]:
treb=treb.reindex(columns=["Nom Cognoms", "Gènere","Fills", "Grup Professional", "Salari mensual", "Nascut_Espanya", "Dia", "Mes", "Any", "Ciutat", "País d'origen"])
treb

,Nom Cognoms,Gènere,Fills,Grup Professional,Salari mensual,Nascut_Espanya,Dia,Mes,Any,Ciutat,País d'origen
DNI,,,,,,,,,,,
28973553Z,Mia Schneider Fischer,Altres,No,Grup A,951 €,False,22,10,1976,Berlín,Alemanya
37399141L,Laura Schneider Fischer,Dona,Sí,Grup B,1.769 €,False,2,2,1958,Berlín,Alemanya
37368317L,Lea Schneider Schneider,Dona,No,Grup B,2.013 €,False,23,10,2005,Berlín,Alemanya
21390098Z,Mia Fischer,Dona,Sí,Grup B,1.557 €,False,11,8,1950,Berlín,Alemanya
44060014R,Jonas Schneider,Home,Sí,Grup D,2.754 €,False,22,11,1985,Berlín,Alemanya
...,...,...,...,...,...,...,...,...,...,...,...
89577876S,Emily Taylor Jones,Dona,No,Grup B,2.033 €,False,28,3,1958,Manchester,Regne Unit
57441590Y,George Brown Jones,Home,Sí,Grup A,1.130 €,False,27,12,1979,Manchester,Regne Unit
58204038A,Olivia Brown Brown,Altres,No,Grup A,1.023 €,False,28,8,1952,Manchester,Regne Unit


<span style="font-size:16px; color:maroon">**Ex.4.**</span>

* <span style="font-size:14px">Crea una ***taula resum*** que permeti veure el sou mig, medià, mínim i màxim per Gènere.</span>




In [146]:
treb.info() #La serie "Salari mensual" és del tipus object, és a dir, un string. 

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, 28973553Z to 82046699J
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Nom Cognoms        1000 non-null   object
 1   Gènere             1000 non-null   object
 2   Fills              1000 non-null   object
 3   Grup Professional  1000 non-null   object
 4   Salari mensual     1000 non-null   object
 5   Nascut_Espanya     1000 non-null   bool  
 6   Dia                1000 non-null   int64 
 7   Mes                1000 non-null   int64 
 8   Any                1000 non-null   int64 
 9   Ciutat             1000 non-null   object
 10  País d'origen      1000 non-null   object
dtypes: bool(1), int64(3), object(7)
memory usage: 86.9+ KB


In [147]:
# Antes de limpiar
print("Antes:", treb["Salari mensual"].dtype, treb["Salari mensual"].head())

Antes: object DNI
28973553Z      951 €
37399141L    1.769 €
37368317L    2.013 €
21390098Z    1.557 €
44060014R    2.754 €
Name: Salari mensual, dtype: object


In [148]:
import numpy as np
treb["Salari mensual"] = (treb["Salari mensual"]
    .astype(str)                    # Convertir tot a string
    .str.replace('€', '', regex=False)  # Treure símbol €
    .str.replace(' ', '', regex=False)  # Treure tots els espais
    .str.replace('.', '', regex=False) # Coma → punt decimal
    .replace('', np.nan)            # String buit → NaN
    .astype(float))                 # Convertir a número


In [149]:
# Després
print("Después:", treb["Salari mensual"].dtype, treb["Salari mensual"].head())

Después: float64 DNI
28973553Z     951.0
37399141L    1769.0
37368317L    2013.0
21390098Z    1557.0
44060014R    2754.0
Name: Salari mensual, dtype: float64


In [150]:
Taula_resum=treb.groupby("Gènere")["Salari mensual"].agg(['mean', 'median', 'max', 'min']).round(2)
Taula_resum

,mean,median,max,min
Gènere,,,,
Altres,1626.59,1545.0,3175.0,703.0
Dona,1469.44,1361.5,3021.0,665.0
Home,1643.25,1531.0,3356.0,737.0
NaN,1568.87,1443.0,2969.0,758.0


* <span style="font-size:14px">Ordena la taula en funció del sou mig.</span>

In [193]:
Taula_resum.sort_values(by=["mean"])   #Potser podriem eliminar la columna NaN

,mean,median,max,min
Gènere,,,,
Dona,1469.44,1361.5,3021.0,665.0
NaN,1568.87,1443.0,2969.0,758.0
Altres,1626.59,1545.0,3175.0,703.0
Home,1643.25,1531.0,3356.0,737.0


<span style="font-size:16px; color:maroon">**Ex.5.**</span>

* <span style="font-size:14px">Crea una taula resum amb el salari mig per gènere (files) i país d'origen (columnes). </span>
* <span style="font-size:14px">Afegeix-hi les mitjanes als marges de la taula.</span>

<span style="font-size:14px; color:salmon ">**(EXTRA)**: Aplica format condicional a la taula per veure en un color més intens els valors més elevats</span>


In [152]:
genpais=treb.groupby(by=["Gènere", "País d'origen"])["Salari mensual"].mean()
genpais.unstack().round(2)

País d'origen,Alemanya,Argentina,Colòmbia,Espanya,França,Itàlia,Marroc,Mèxic,Portugal,Regne Unit
Gènere,,,,,,,,,,
Altres,951.00,1141.00,1030.00,1706.18,NaN,1423.00,1365.00,1372.0,1765.00,1921.00
Dona,1804.31,1291.80,1497.75,1460.16,1566.47,1247.18,1405.21,1517.8,1488.55,1489.46
Home,2067.43,1583.29,1554.67,1682.11,1389.25,1672.88,1531.00,1625.0,1497.00,1162.56
NaN,1931.50,1135.67,1252.00,1597.14,1573.00,1316.00,1365.50,1583.0,1553.50,1758.00


In [153]:
treb.pivot_table("Salari mensual", index="Gènere", columns="País d'origen", observed=True).round(2)

País d'origen,Alemanya,Argentina,Colòmbia,Espanya,França,Itàlia,Marroc,Mèxic,Portugal,Regne Unit
Gènere,,,,,,,,,,
Altres,951.00,1141.00,1030.00,1706.18,NaN,1423.00,1365.00,1372.0,1765.00,1921.00
Dona,1804.31,1291.80,1497.75,1460.16,1566.47,1247.18,1405.21,1517.8,1488.55,1489.46
Home,2067.43,1583.29,1554.67,1682.11,1389.25,1672.88,1531.00,1625.0,1497.00,1162.56
NaN,1931.50,1135.67,1252.00,1597.14,1573.00,1316.00,1365.50,1583.0,1553.50,1758.00


* <span style="font-size:14px">Afegeix-hi les mitjanes als marges de la taula.</span>

<span style="font-size:14px; color:salmon ">**(EXTRA)**: Aplica format condicional a la taula per veure en un color més intens els valors més elevats</span>

<span style="font-size:14px">**Referencia pel canvi de color:** 

*  <span style="font-size:14px">[
DataFrames con Estilo: Formatos condicionales, colores y más.](https://www.youtube.com/watch?v=swzuFTMKAwk) --> **Tech Portal Formación** </span>

In [154]:
Treb_pivot=treb.pivot_table("Salari mensual", 
                 index="Gènere", 
                 columns="País d'origen", 
                 margins=True, 
                 margins_name='Mitjana', 
                 aggfunc='mean', 
                 dropna=True).round(3)
Treb_pivot.style.format(
    precision=3
    ).apply(
        lambda x: ['color: coral' 
                   if v==x.max() else '' 
                   for v in x])

País d'origen,Alemanya,Argentina,Colòmbia,Espanya,França,Itàlia,Marroc,Mèxic,Portugal,Regne Unit,Mitjana
Gènere,,,,,,,,,,,
Altres,951.000,1141.000,1030.000,1706.176,nan,1423.000,1365.000,1372.000,1765.000,1921.000,1626.588
Dona,1804.308,1291.800,1497.750,1460.164,1566.471,1247.182,1405.214,1517.800,1488.545,1489.462,1469.443
Home,2067.429,1583.294,1554.667,1682.106,1389.250,1672.875,1531.000,1625.000,1497.000,1162.556,1643.249
NaN,1931.500,1135.667,1252.000,1597.140,1573.000,1316.000,1365.500,1583.000,1553.500,1758.000,1568.867
Mitjana,1858.348,1431.677,1489.132,1582.163,1465.357,1421.174,1441.690,1559.081,1527.226,1448.333,1561.462


In [155]:
Treb_pivot.style.format(precision=3
                        ).apply(lambda x: 
                                ['background-color:orange'
                                 if v==x.max() else '' 
                                 for v in x])

País d'origen,Alemanya,Argentina,Colòmbia,Espanya,França,Itàlia,Marroc,Mèxic,Portugal,Regne Unit,Mitjana
Gènere,,,,,,,,,,,
Altres,951.000,1141.000,1030.000,1706.176,nan,1423.000,1365.000,1372.000,1765.000,1921.000,1626.588
Dona,1804.308,1291.800,1497.750,1460.164,1566.471,1247.182,1405.214,1517.800,1488.545,1489.462,1469.443
Home,2067.429,1583.294,1554.667,1682.106,1389.250,1672.875,1531.000,1625.000,1497.000,1162.556,1643.249
NaN,1931.500,1135.667,1252.000,1597.140,1573.000,1316.000,1365.500,1583.000,1553.500,1758.000,1568.867
Mitjana,1858.348,1431.677,1489.132,1582.163,1465.357,1421.174,1441.690,1559.081,1527.226,1448.333,1561.462


<span style="font-size:16px; color: Maroon">**Ex.6.**</span>

* <span style="font-size:14px">Crea una columna nova que sigui la data de naixament en format Datetime a partir de les columnes dia, mes i any. Utilitzant aquesta columna crea una funció que donada una data, et calculi l'edat actual a dia d'avui.</span>

In [156]:
Nulls=treb.isnull().any() #Busquem a veure si hi ha algun valor null (NaN) en les columnes del df
Nulls
#No hi ha valors nulls

Nom Cognoms          False
Gènere               False
Fills                False
Grup Professional    False
Salari mensual       False
Nascut_Espanya       False
Dia                  False
Mes                  False
Any                  False
Ciutat               False
País d'origen        False
dtype: bool

In [157]:
# En el nostre cas en comptes de valors NaN potser tenim celles buides i per tant hem de fer el següent:
import numpy as np
# Reemplazar cadenas vacías por NaN y luego contar
treb.replace(r'^\s*$', np.nan, regex=True).isnull().sum()  #La sortida després d'execució dona zero en totes les columnes. 


Nom Cognoms          0
Gènere               0
Fills                0
Grup Professional    0
Salari mensual       0
Nascut_Espanya       0
Dia                  0
Mes                  0
Any                  0
Ciutat               0
País d'origen        0
dtype: int64

In [158]:
# Renombrar columnas
temp_df = treb[['Dia', 'Mes', 'Any']].rename(columns={'Dia': 'day', 'Mes': 'month', 'Any': 'year'})
# Convertir
treb['Data Naixement'] = pd.to_datetime(temp_df)
treb


,Nom Cognoms,Gènere,Fills,Grup Professional,Salari mensual,Nascut_Espanya,Dia,Mes,Any,Ciutat,País d'origen,Data Naixement
DNI,,,,,,,,,,,,
28973553Z,Mia Schneider Fischer,Altres,No,Grup A,951.0,False,22,10,1976,Berlín,Alemanya,1976-10-22
37399141L,Laura Schneider Fischer,Dona,Sí,Grup B,1769.0,False,2,2,1958,Berlín,Alemanya,1958-02-02
37368317L,Lea Schneider Schneider,Dona,No,Grup B,2013.0,False,23,10,2005,Berlín,Alemanya,2005-10-23
21390098Z,Mia Fischer,Dona,Sí,Grup B,1557.0,False,11,8,1950,Berlín,Alemanya,1950-08-11
44060014R,Jonas Schneider,Home,Sí,Grup D,2754.0,False,22,11,1985,Berlín,Alemanya,1985-11-22
...,...,...,...,...,...,...,...,...,...,...,...,...
89577876S,Emily Taylor Jones,Dona,No,Grup B,2033.0,False,28,3,1958,Manchester,Regne Unit,1958-03-28
57441590Y,George Brown Jones,Home,Sí,Grup A,1130.0,False,27,12,1979,Manchester,Regne Unit,1979-12-27
58204038A,Olivia Brown Brown,Altres,No,Grup A,1023.0,False,28,8,1952,Manchester,Regne Unit,1952-08-28


* <span style="font-size:14px">Utilitza la funció que acabes de crear per generar una columna nova al DataFrame amb l'edat actual.</span>

In [159]:
treb.drop(columns=["Dia", "Mes", "Any"])

,Nom Cognoms,Gènere,Fills,Grup Professional,Salari mensual,Nascut_Espanya,Ciutat,País d'origen,Data Naixement
DNI,,,,,,,,,
28973553Z,Mia Schneider Fischer,Altres,No,Grup A,951.0,False,Berlín,Alemanya,1976-10-22
37399141L,Laura Schneider Fischer,Dona,Sí,Grup B,1769.0,False,Berlín,Alemanya,1958-02-02
37368317L,Lea Schneider Schneider,Dona,No,Grup B,2013.0,False,Berlín,Alemanya,2005-10-23
21390098Z,Mia Fischer,Dona,Sí,Grup B,1557.0,False,Berlín,Alemanya,1950-08-11
44060014R,Jonas Schneider,Home,Sí,Grup D,2754.0,False,Berlín,Alemanya,1985-11-22
...,...,...,...,...,...,...,...,...,...
89577876S,Emily Taylor Jones,Dona,No,Grup B,2033.0,False,Manchester,Regne Unit,1958-03-28
57441590Y,George Brown Jones,Home,Sí,Grup A,1130.0,False,Manchester,Regne Unit,1979-12-27
58204038A,Olivia Brown Brown,Altres,No,Grup A,1023.0,False,Manchester,Regne Unit,1952-08-28


In [160]:
treb=treb.reindex(columns=["Nom Cognoms", "Data Naixement","Gènere","Fills", "Grup Professional", "Salari mensual", "Nascut_Espanya", "Ciutat", "País d'origen"])
treb

,Nom Cognoms,Data Naixement,Gènere,Fills,Grup Professional,Salari mensual,Nascut_Espanya,Ciutat,País d'origen
DNI,,,,,,,,,
28973553Z,Mia Schneider Fischer,1976-10-22,Altres,No,Grup A,951.0,False,Berlín,Alemanya
37399141L,Laura Schneider Fischer,1958-02-02,Dona,Sí,Grup B,1769.0,False,Berlín,Alemanya
37368317L,Lea Schneider Schneider,2005-10-23,Dona,No,Grup B,2013.0,False,Berlín,Alemanya
21390098Z,Mia Fischer,1950-08-11,Dona,Sí,Grup B,1557.0,False,Berlín,Alemanya
44060014R,Jonas Schneider,1985-11-22,Home,Sí,Grup D,2754.0,False,Berlín,Alemanya
...,...,...,...,...,...,...,...,...,...
89577876S,Emily Taylor Jones,1958-03-28,Dona,No,Grup B,2033.0,False,Manchester,Regne Unit
57441590Y,George Brown Jones,1979-12-27,Home,Sí,Grup A,1130.0,False,Manchester,Regne Unit
58204038A,Olivia Brown Brown,1952-08-28,Altres,No,Grup A,1023.0,False,Manchester,Regne Unit


In [161]:
treb["Edat"]=treb["Data Naixement"].apply(lambda x: 2025 - x.year)
treb 

,Nom Cognoms,Data Naixement,Gènere,Fills,Grup Professional,Salari mensual,Nascut_Espanya,Ciutat,País d'origen,Edat
DNI,,,,,,,,,,
28973553Z,Mia Schneider Fischer,1976-10-22,Altres,No,Grup A,951.0,False,Berlín,Alemanya,49
37399141L,Laura Schneider Fischer,1958-02-02,Dona,Sí,Grup B,1769.0,False,Berlín,Alemanya,67
37368317L,Lea Schneider Schneider,2005-10-23,Dona,No,Grup B,2013.0,False,Berlín,Alemanya,20
21390098Z,Mia Fischer,1950-08-11,Dona,Sí,Grup B,1557.0,False,Berlín,Alemanya,75
44060014R,Jonas Schneider,1985-11-22,Home,Sí,Grup D,2754.0,False,Berlín,Alemanya,40
...,...,...,...,...,...,...,...,...,...,...
89577876S,Emily Taylor Jones,1958-03-28,Dona,No,Grup B,2033.0,False,Manchester,Regne Unit,67
57441590Y,George Brown Jones,1979-12-27,Home,Sí,Grup A,1130.0,False,Manchester,Regne Unit,46
58204038A,Olivia Brown Brown,1952-08-28,Altres,No,Grup A,1023.0,False,Manchester,Regne Unit,73


* <span style="font-size:16px">Com es pot fer perquè sigui un calcul automàtic? És a dir, que agafi la data d'avui (actual) i retorni l'edat del treballador/-a⁉️

* <span style="font-size:14px; color: lawngreen">Proposta de solució: 
        
        * Ajustar l'edat a el dia de naixement, el mes i l'any. Exercici a fer més endavant‼️

In [162]:
from datetime import datetime
treb["Edat"]=treb["Data Naixement"].apply(lambda x: datetime.now().year - x.year) #Aquesta funció agafa l'any actual i el resta a l'any de naixement per obtenir l'edat.
treb
#lambda x: datetime.now().year - x.year --> datetime.now() retorna la data i hora actual, i .year extreu l'any d'aquella data. 
#Al restar l'any actual amb l'any de naixement, obtenim l'edat actual del client o el que sigui.

,Nom Cognoms,Data Naixement,Gènere,Fills,Grup Professional,Salari mensual,Nascut_Espanya,Ciutat,País d'origen,Edat
DNI,,,,,,,,,,
28973553Z,Mia Schneider Fischer,1976-10-22,Altres,No,Grup A,951.0,False,Berlín,Alemanya,50
37399141L,Laura Schneider Fischer,1958-02-02,Dona,Sí,Grup B,1769.0,False,Berlín,Alemanya,68
37368317L,Lea Schneider Schneider,2005-10-23,Dona,No,Grup B,2013.0,False,Berlín,Alemanya,21
21390098Z,Mia Fischer,1950-08-11,Dona,Sí,Grup B,1557.0,False,Berlín,Alemanya,76
44060014R,Jonas Schneider,1985-11-22,Home,Sí,Grup D,2754.0,False,Berlín,Alemanya,41
...,...,...,...,...,...,...,...,...,...,...
89577876S,Emily Taylor Jones,1958-03-28,Dona,No,Grup B,2033.0,False,Manchester,Regne Unit,68
57441590Y,George Brown Jones,1979-12-27,Home,Sí,Grup A,1130.0,False,Manchester,Regne Unit,47
58204038A,Olivia Brown Brown,1952-08-28,Altres,No,Grup A,1023.0,False,Manchester,Regne Unit,74



<span style="font-size:40px; color: paleturquoise">***Nivell 2*** </span>

<span style="font-size:16px; color: maroon">**Ex.1.**</span>

* <span style="font-size:14px">Utilitzant el següent DataFrame, adjunta la columna "Increment" al dataframe del nivell anterior.
* <span style="font-size:14px">Actualitza la columna salari en funció dels percentatges que s'adjunten. No modifiquis manualment els increments, escriu codi Python per fer les conversions necessàries. <br><br>
<span style="font-size:14 px; color: darkviolet"> <center> **df_increment** = pd.DataFrame({"Grup":["Grup A","Grup B","Grup C", "Grup D" ] ,<br> **"Increment"**:
["5%","3,5%","2%","8%"]}) </center>

In [163]:
df_increment= pd.DataFrame({"Grup":["Grup A","Grup B","Grup C", "Grup D" ] ,"Increment":
["5%","3,5%","2%","8%"]})
df_increment

,Grup,Increment
0,Grup A,5%
1,Grup B,"3,5%"
2,Grup C,2%
3,Grup D,8%


In [164]:
# S'ha de fer un merge per adjuntar la columna "Increment" al df treb. 
# Per això, primer hem de canviar el nom de la columna "Grup" a "Grup Professional" per tal que coincideixi amb el nom de la columna del df treb.
df_increment.rename(columns={"Grup":"Grup Professional"}, inplace=True)
df_increment

,Grup Professional,Increment
0,Grup A,5%
1,Grup B,"3,5%"
2,Grup C,2%
3,Grup D,8%


In [165]:
# MERGE
treb=pd.merge(treb, df_increment, on="Grup Professional", how="left")
treb=treb.reindex(columns=["Nom Cognoms", "Data Naixement","Edat", "Gènere","Fills", "Grup Professional", "Increment", "Salari mensual", "Nascut_Espanya", "Ciutat", "País d'origen"])
treb

,Nom Cognoms,Data Naixement,Edat,Gènere,Fills,Grup Professional,Increment,Salari mensual,Nascut_Espanya,Ciutat,País d'origen
0,Mia Schneider Fischer,1976-10-22,50,Altres,No,Grup A,5%,951.0,False,Berlín,Alemanya
1,Laura Schneider Fischer,1958-02-02,68,Dona,Sí,Grup B,"3,5%",1769.0,False,Berlín,Alemanya
2,Lea Schneider Schneider,2005-10-23,21,Dona,No,Grup B,"3,5%",2013.0,False,Berlín,Alemanya
3,Mia Fischer,1950-08-11,76,Dona,Sí,Grup B,"3,5%",1557.0,False,Berlín,Alemanya
4,Jonas Schneider,1985-11-22,41,Home,Sí,Grup D,8%,2754.0,False,Berlín,Alemanya
...,...,...,...,...,...,...,...,...,...,...,...
995,Emily Taylor Jones,1958-03-28,68,Dona,No,Grup B,"3,5%",2033.0,False,Manchester,Regne Unit
996,George Brown Jones,1979-12-27,47,Home,Sí,Grup A,5%,1130.0,False,Manchester,Regne Unit
997,Olivia Brown Brown,1952-08-28,74,Altres,No,Grup A,5%,1023.0,False,Manchester,Regne Unit
998,Isla Jones Brown,1999-03-28,27,Dona,No,Grup A,5%,1197.0,False,Manchester,Regne Unit


In [166]:
#Comprovem de nou que no hi ha valors nulls a la columna "Increment" després de fer el merge.
treb["Increment"].isnull().sum() #La sortida és 0, per tant no hi ha valors nulls a la columna "Increment".

np.int64(0)

In [167]:
# De quin tipus és la columna "Increment"? Podem treballar amb ella o abans l'hem de transformar a tipus numèric?
treb["Increment"].dtype #La columna "Increment" és del tipus object.

dtype('O')

In [168]:
#Transformem a tipus numèric:
treb["Increment"] = (treb["Increment"]
    .astype(str)                    # Convertir tot a string
    .str.replace('%', '', regex=False)  # Treure símbol %
    .str.replace(',', '.', regex=False)  # Coma → punt decimal
    .replace('', np.nan)            # String buit → NaN
    .astype(float))                 # Convertir a número

In [169]:
#Comprovem que és del tipus numèric després de la transformació:
treb["Increment"].dtype

dtype('float64')

* <span style="font-size:14px">Actualitza la columna salari en funció dels percentatges que s'adjunten. No modifiquis manualment els increments, escriu codi Python per fer les conversions necessàries. <br><br>

In [170]:
treb["Nou salari"]=treb.apply(
    lambda row: row["Salari mensual"] * 1.05 if row["Grup Professional"] == "Grup A"
                else row["Salari mensual"] * 1.035 if row["Grup Professional"] == "Grup B"
                else row["Salari mensual"] * 1.02 if row["Grup Professional"] == "Grup C"
                else row["Salari mensual"],
    axis=1
).round(2)
treb

,Nom Cognoms,Data Naixement,Edat,Gènere,Fills,Grup Professional,Increment,Salari mensual,Nascut_Espanya,Ciutat,País d'origen,Nou salari
0,Mia Schneider Fischer,1976-10-22,50,Altres,No,Grup A,5.0,951.0,False,Berlín,Alemanya,998.55
1,Laura Schneider Fischer,1958-02-02,68,Dona,Sí,Grup B,3.5,1769.0,False,Berlín,Alemanya,1830.92
2,Lea Schneider Schneider,2005-10-23,21,Dona,No,Grup B,3.5,2013.0,False,Berlín,Alemanya,2083.46
3,Mia Fischer,1950-08-11,76,Dona,Sí,Grup B,3.5,1557.0,False,Berlín,Alemanya,1611.50
4,Jonas Schneider,1985-11-22,41,Home,Sí,Grup D,8.0,2754.0,False,Berlín,Alemanya,2754.00
...,...,...,...,...,...,...,...,...,...,...,...,...
995,Emily Taylor Jones,1958-03-28,68,Dona,No,Grup B,3.5,2033.0,False,Manchester,Regne Unit,2104.15
996,George Brown Jones,1979-12-27,47,Home,Sí,Grup A,5.0,1130.0,False,Manchester,Regne Unit,1186.50
997,Olivia Brown Brown,1952-08-28,74,Altres,No,Grup A,5.0,1023.0,False,Manchester,Regne Unit,1074.15
998,Isla Jones Brown,1999-03-28,27,Dona,No,Grup A,5.0,1197.0,False,Manchester,Regne Unit,1256.85


In [171]:
""""
treb["Salari mensual"] = treb.apply(
    lambda row: row["Salari mensual"] * 1.05 if row["Grup Professional"] == "Grup A"
                else row["Salari mensual"] * 1.035 if row["Grup Professional"] == "Grup B"
                else row["Salari mensual"] * 1.02 if row["Grup Professional"] == "Grup C"
                else row["Salari mensual"],
    axis=1
).round(2)
treb
"""

'"\ntreb["Salari mensual"] = treb.apply(\n    lambda row: row["Salari mensual"] * 1.05 if row["Grup Professional"] == "Grup A"\n                else row["Salari mensual"] * 1.035 if row["Grup Professional"] == "Grup B"\n                else row["Salari mensual"] * 1.02 if row["Grup Professional"] == "Grup C"\n                else row["Salari mensual"],\n    axis=1\n).round(2)\ntreb\n'

<span style="font-size:16px; color: maroon">**Ex.2.**</span>

* <span style="font-size:14px">Utilitzant un bucle, exporta en 4 fitxers (format .xlsx o .csv) les dades de cada Grup Professional.<br><font color="plum"> **Per exemple**:</font> "dades_GrupA.xlsx" , "dades_GrupB.xlsx" ... <br>
* <span style="font-size:14px">Exporta un 5è DataFrame en format .xlsx o .csv que contingui quants treballadors hi ha per cada Grup Professional, quin és el seu sou mig i quina és la seva edat mediana.

In [172]:
# Exportar en un excel cada grup professional a un full diferent:
with pd.ExcelWriter(r"C:\Users\dioul\OneDrive\Escritorio\Fonaments D'anàlisi de dades (FADD)\Curs d'especialització Data Analys\Python\Sprint _10\Exportats_Nivell_2\treballadors_2025.xlsx") as writer:
    for grup, df_grup in treb.groupby("Grup Professional"): 
        df_grup.to_excel(writer, sheet_name=grup, index=False)   #index=False per no exportar la columna de l'índex al fitxer excel.

#Què s'ha fet aquí?
#1) with pd.ExcelWriter(...) as writer: --> Aquesta línia crea un objecte ExcelWriter que ens permet escriure múltiples fulls en un mateix fitxer Excel. El camí del fitxer es passa com a argument.
#2) for grup, df_grup in treb.groupby("Grup Professional"): --> Aquesta línia itera sobre cada grup professional del DataFrame treb. La funció groupby agrupa
#el DataFrame per la columna "Grup Professional" i retorna un iterador que produeix parells de (grup, df_grup), on grup és el nom del grup professional i df_grup és un DataFrame que conté només les files corresponents a aquell grup.
#3) df_grup.to_excel(writer, sheet_name=grup, index=False) --> Aquesta línia escriu el DataFrame df_grup en un full de l'arxiu
    

In [173]:
#Alternativa per exportar a excel cada grup professional a un fitxer diferent: "dades_GrupA.xlsx", etc
with pd.ExcelWriter(r"C:\Users\dioul\OneDrive\Escritorio\Fonaments D'anàlisi de dades (FADD)\Curs d'especialització Data Analys\Python\Sprint _10\Exportats_Nivell_2\dades_GrupA.xlsx") as writer:
    treb[treb["Grup Professional"]=="Grup A"].to_excel(writer, index=False)

#Fem el mateix per la resta de grups.

* <span style="font-size:14px">Exporta un 5è DataFrame en format .xlsx o .csv que contingui quants treballadors hi ha per cada Grup Professional, quin és el seu sou mig i quina és la seva edat mediana.

In [174]:
# Calcular la mitjana de salari per Grup i Edat:
resum = treb.groupby(["Grup Professional", "Edat"])["Salari mensual"].mean().reset_index().round(2)
resum.columns = ["Grup Professional", "Edat", "Salari mensual mitjà"]

# Exportar a una fulla excel
resum.to_excel(r"C:\Users\dioul\OneDrive\Escritorio\Fonaments D'anàlisi de dades (FADD)\Curs d'especialització Data Analys\Python\Sprint _10\Exportats_Nivell_2\Taula_Resum.xlsx", index=False)


In [175]:
#També podem crear la taula com a Pivot Table i exportar-les dades de forma més estructurada (segons els nostres criteris):



<span style="font-size:40px; color: paleturquoise">***Nivell 3*** </span>

<span style="font-size:16px; color: crimson">***El nivell 3 d’aquest sprint és totalment diferent a d’altres sprints que has fet fins ara, ja que són exercicis més abstractes que requereixen barallar-s’hi bastant.<br> No continuen amb el mateix dataset dels nivells anteriors, sinó que et plantegen dues situacions noves totalment diferents entre elles.***

<span style="font-size:16px; color: maroon">**1.
Crea una funció que prengui un dataframe com a paràmetre d'entrada.<br><br>
La funció ha de crear (i exportar) un gràfic automàticament per a cada columna del dataframe**. *Per exemple:*</span>

* <span style="font-size:14px">un histograma/boxplot si la variable és numèrica
* <span style="font-size:14px">unes barres dels valors més freqüents si és categòrica
* <span style="font-size:14px">unes barres dels anys més freqüents si la dada està en format data.

<span style="font-size:14px">La idea és crear una funció que funcioni per <font color="gold"> **qualsevol**</font> dataframe, no només amb el que hem treballat fins ara.</span>

<span style="font-size:14px">Mostra el resultat de la funció en algun dels datasets d’exemple que conté el paquet seaborn. Per exemple, iris, penguins o titanic.</span>

<span style="font-size:14px; color:maroon">**Tingues en consideració que en el següent sprint treballaràs exclusivament amb gràfics. L’objectiu d’aquest exercici no és crear gràfics molt elaborats, sinó resoldre una necessitat de manera ràpida i automàtica.**

In [176]:
import seaborn as sns

In [177]:
titanic=sns.load_dataset("titanic")
titanic

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True


In [178]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [179]:
titanic.shape

(891, 15)

In [180]:
titanic.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [181]:
col_nu=titanic.select_dtypes(include="number").columns
col_nu

Index(['survived', 'pclass', 'age', 'sibsp', 'parch', 'fare'], dtype='object')

In [182]:
col_cat=titanic.select_dtypes(include='object').columns
col_cat

Index(['sex', 'embarked', 'who', 'embark_town', 'alive'], dtype='object')

In [183]:

def graficos_automaticos(df):
    import pandas as pd
    import matplotlib.pyplot as plt
    from pathlib import Path

    carpeta = Path(r"C:\Users\dioul\OneDrive\Escritorio\Fonaments D'anàlisi de dades (FADD)\Curs d'especialització Data Analys\Python\Sprint _10\Gràfics_automatitzats")
    carpeta.mkdir(exist_ok=True)

    for nom_columna in df.columns:
        serie = df[nom_columna]
        valors_unics = serie.dropna().unique()
        n_unics = len(valors_unics)

        plt.figure(figsize=(8, 6))

        if serie.dtype in ["int64", "float64"] and n_unics < 4:
            # numèrica però amb pocs valors únics: categórica (0,1)
            recompte = serie.value_counts().sort_index()                #El sort_index m'ordena els valors de major a menor.
            recompte.plot(kind="bar")
            plt.title(f"{nom_columna} (Categorica - binaria -)")        # A la gràfica li posem com a títol el nom de la columna i entre parentesis que és Cdel tipus CATEGORICA.
        elif serie.dtype in ["int64", "float64"]:
            # numérica “real”
            if len(df) > 10:
                serie.plot.box()
                plt.title(f"{nom_columna} - boxplot")
            else:
                plt.hist(serie.dropna(), bins=10, edgecolor="black")       #bins=10 ens diu que són valors binaris fins a 10. 
                plt.title(f"{nom_columna} - histograma")
        else:
            # texto/categórica
            conteo = serie.value_counts().head(10)
            conteo.plot(kind="bar")
            plt.title(f"{nom_columna} - més freqüents")
            plt.xticks(rotation=45)

        plt.ylabel("Quantitat")
        plt.tight_layout()

        # usar SOLO el nombre de la columna para el fichero
        nombre_fichero = f"{str(nom_columna)}.png"
        plt.savefig(carpeta / nombre_fichero)
        plt.close()

    print(f"Okay, ja ho tenim això: {len(df.columns)} gràfics en la carpeta 'Gràfics_automatitzats'")



In [184]:
graficos_automaticos(titanic)

Okay, ja ho tenim això: 15 gràfics en la carpeta 'Gràfics_automatitzats'


<span style="font-size:16px; color:plum">Què és el que saben del dataframe que volem analitzar? 

* <span style="font-size:14px">Tenim els següents tipus de dades: **int, float, object, i categòriques**

* <span style="font-size:14px">Hi ha *15 columnes*. La llarga del df és de *891 registres*. 

* <span style="font-size:14px">Hi ha 2 columnes amb valors nulls. Però si analitzem la sortida del mètode .info(), observem que hi ha 4 columnes amb *Missing values* — age, embarked, deck, i embark_town. 

<span style="font-size:16px; color:maroon">***2. Carrega l'arxiu matriu_distancies.xlsx a pandas, de manera que els noms de files i els noms de columnes siguin els de les ciutats. Borra "Las Palmas de Gran Canaria" i "Palma" perquè poguem fer el trajecte en cotxe.***</span>

<span style="font-size:14px"> **Font**: [Mejores Rutas](https://espana.mejoresrutas.com/tabla-de-distancias/es/)

<span style="font-size:14px">Ens interessa visitar totes les ciutats principals d'Espanya recorrent la mínima distància possible.</span>

<span style="font-size:14px; color:aquamarine">***No cal que ho facis de forma òptima, ens interessa que desenvolupis una solució raonable utilitzant les eines que tens actualment.***</span>

<span style="font-size:14px">Per exemple, una aproximació senzilla (que no òptima) seria anant sempre a la ciutat més propera que no haguem visitat encara

<span style="font-size:14px">Fes una **funció** que donada la matriu de distàncies i la ciutat d'origen, faci una proposta de ruta que sigui el més curta possible que puguis, retornant una llista amb l'ordre de visita. Dóna també la distància total recorreguda.</span>

<span style="font-size:14px; color:salmon ">**(EXTRA)** Des de quina ciutat la ruta seria més curta amb l'algoritme plantejat

In [2]:
from pathlib import Path
import pandas as pd
BASE_DIR=Path.cwd()
ruta=BASE_DIR/'es.xlsx'                  #Hem de instal·lar 'xlrd' per obrir versions actigues d'Excel. L'arxiu és xls del 97-2003
matriu_dist=pd.read_excel(ruta, header=0, index_col=0)
matriu_dist

,Barcelona,Valencia,Sevilla,Zaragoza,Málaga,Murcia,Palma,Las Palmas de Gran Canaria,Bilbao,Alicante,Córdoba,Valladolid,Vigo,Gijón,Hospitalet de Llobregat
Distancia (kilómetros),,,,,,,,,,,,,,,
Barcelona,NaN,350.0,999.0,310.0,953.0,567.0,715.0,2519.0,609.0,516.0,863.0,727.0,1142.0,874.0,13.0
Valencia,350.0,NaN,658.0,310.0,604.0,218.0,366.0,2179.0,612.0,166.0,523.0,566.0,946.0,822.0,342.0
Sevilla,998.0,659.0,NaN,841.0,210.0,528.0,951.0,1523.0,859.0,601.0,143.0,581.0,773.0,796.0,990.0
Zaragoza,309.0,310.0,842.0,NaN,834.0,526.0,675.0,2358.0,303.0,475.0,701.0,421.0,836.0,568.0,301.0
Málaga,953.0,607.0,210.0,834.0,NaN,400.0,823.0,1731.0,923.0,473.0,159.0,739.0,979.0,1002.0,946.0
Murcia,566.0,220.0,527.0,526.0,400.0,NaN,431.0,2049.0,781.0,82.0,434.0,607.0,986.0,863.0,559.0
Palma,714.0,369.0,950.0,675.0,823.0,431.0,NaN,2472.0,977.0,353.0,874.0,918.0,1298.0,1174.0,707.0
Las Palmas de Gran Canaria,2520.0,2181.0,1525.0,2360.0,1732.0,2050.0,2473.0,NaN,2377.0,2123.0,1665.0,2100.0,2227.0,2314.0,2513.0
Bilbao,609.0,611.0,859.0,303.0,922.0,782.0,976.0,2375.0,NaN,776.0,789.0,279.0,666.0,265.0,601.0


In [3]:
mtrx_neta=matriu_dist.dropna(axis=0, how='all').copy()    #.copy() elimina el Warning que ens surt després quan fem el .drop(inplace=True)
mtrx_neta

,Barcelona,Valencia,Sevilla,Zaragoza,Málaga,Murcia,Palma,Las Palmas de Gran Canaria,Bilbao,Alicante,Córdoba,Valladolid,Vigo,Gijón,Hospitalet de Llobregat
Distancia (kilómetros),,,,,,,,,,,,,,,
Barcelona,NaN,350.0,999.0,310.0,953.0,567.0,715.0,2519.0,609.0,516.0,863.0,727.0,1142.0,874.0,13.0
Valencia,350.0,NaN,658.0,310.0,604.0,218.0,366.0,2179.0,612.0,166.0,523.0,566.0,946.0,822.0,342.0
Sevilla,998.0,659.0,NaN,841.0,210.0,528.0,951.0,1523.0,859.0,601.0,143.0,581.0,773.0,796.0,990.0
Zaragoza,309.0,310.0,842.0,NaN,834.0,526.0,675.0,2358.0,303.0,475.0,701.0,421.0,836.0,568.0,301.0
Málaga,953.0,607.0,210.0,834.0,NaN,400.0,823.0,1731.0,923.0,473.0,159.0,739.0,979.0,1002.0,946.0
Murcia,566.0,220.0,527.0,526.0,400.0,NaN,431.0,2049.0,781.0,82.0,434.0,607.0,986.0,863.0,559.0
Palma,714.0,369.0,950.0,675.0,823.0,431.0,NaN,2472.0,977.0,353.0,874.0,918.0,1298.0,1174.0,707.0
Las Palmas de Gran Canaria,2520.0,2181.0,1525.0,2360.0,1732.0,2050.0,2473.0,NaN,2377.0,2123.0,1665.0,2100.0,2227.0,2314.0,2513.0
Bilbao,609.0,611.0,859.0,303.0,922.0,782.0,976.0,2375.0,NaN,776.0,789.0,279.0,666.0,265.0,601.0


In [4]:
mtrx_neta.drop(columns={'Las Palmas de Gran Canaria', 'Palma'}, 
               index={'Las Palmas de Gran Canaria', 'Palma'}, 
               inplace=True)

In [5]:
mtrx_neta

,Barcelona,Valencia,Sevilla,Zaragoza,Málaga,Murcia,Bilbao,Alicante,Córdoba,Valladolid,Vigo,Gijón,Hospitalet de Llobregat
Distancia (kilómetros),,,,,,,,,,,,,
Barcelona,NaN,350.0,999.0,310.0,953.0,567.0,609.0,516.0,863.0,727.0,1142.0,874.0,13.0
Valencia,350.0,NaN,658.0,310.0,604.0,218.0,612.0,166.0,523.0,566.0,946.0,822.0,342.0
Sevilla,998.0,659.0,NaN,841.0,210.0,528.0,859.0,601.0,143.0,581.0,773.0,796.0,990.0
Zaragoza,309.0,310.0,842.0,NaN,834.0,526.0,303.0,475.0,701.0,421.0,836.0,568.0,301.0
Málaga,953.0,607.0,210.0,834.0,NaN,400.0,923.0,473.0,159.0,739.0,979.0,1002.0,946.0
Murcia,566.0,220.0,527.0,526.0,400.0,NaN,781.0,82.0,434.0,607.0,986.0,863.0,559.0
Bilbao,609.0,611.0,859.0,303.0,922.0,782.0,NaN,776.0,789.0,279.0,666.0,265.0,601.0
Alicante,514.0,168.0,600.0,474.0,473.0,82.0,776.0,NaN,507.0,631.0,1011.0,887.0,506.0
Córdoba,860.0,521.0,143.0,698.0,157.0,433.0,787.0,506.0,NaN,604.0,830.0,860.0,853.0


<span style="font-size:14px">Fes una **funció** que donada la matriu de distàncies i la ciutat d'origen, faci una proposta de ruta que sigui el més curta possible que puguis, retornant una llista amb l'ordre de visita. Dóna també la distància total recorreguda.</span>

<span style="font-size:14px; color:salmon ">**(EXTRA)** Des de quina ciutat la ruta seria més curta amb l'algoritme plantejat

In [6]:
def info_distancies(matriu, origen):
    if origen not in matriu.index:
        raise ValueError(f"La ciutat '{origen}' no existeix a la matriu.")
    dist_origen = matriu[origen]
    dist_ordenada = dist_origen.sort_values(na_position='first')
    diferencia = dist_ordenada.diff()
    df = pd.DataFrame({
        "Distància des d'origen": dist_ordenada,
        "Diferència": diferencia
    })

    return df


In [12]:
ciutat=input("Entra una ciutat d'origen: ")
df_resultat = info_distancies(mtrx_neta, ciutat)
df_resultat


,Distància des d'origen,Diferència
Distancia (kilómetros),,
Barcelona,NaN,NaN
Hospitalet de Llobregat,10.0,NaN
Zaragoza,309.0,299.0
Valencia,350.0,41.0
Alicante,514.0,164.0
Murcia,566.0,52.0
Bilbao,609.0,43.0
Valladolid,728.0,119.0
Córdoba,860.0,132.0


<span style="color:peru; font-size:14px">**Què pasa si l'usuari vol comparar més d'una ciutat?**

<span style="font-size:12px">Donada una matriu de distàncies i una llista de ciutats d'origen, retorna un DataFrame amb:
* <span style="font-size:12px">Ciutats com a index
* <span style="font-size:12px">Per cada ciutat d'origen:
    * <span style="font-size:12px">Distància ordenada
    * <span style="font-size:12px">Diferència 

In [ ]:
def info_distancies_multi(matriu, ciutats_origen):
    for ciutat in ciutats_origen:
        if ciutat not in matriu:
            raise ValueError(f"La ciutat '{ciutat}' no existeix a la matriu.")

    # DataFrame final buit
    df_final = pd.DataFrame(index=matriu.index)

    # Per cada ciutat d'origen, calculem distàncies i diferències
    for origen in ciutats_origen:
        dist_origen = matriu[origen]

        # Ordenar distàncies
        dist_ordenada = dist_origen.sort_values(na_position='first')

        diferencia = dist_ordenada.diff()
        df_final[f"Distància des de {origen}"] = dist_ordenada
        #df_final[f"Diferència {origen}"] = diferencia

    return df_final


In [ ]:
ciutat = input("Entra ciutat o ciutats d'origen (separades per comes): ")
llista_ciutats = [c.strip() for c in ciutat.split(",")]

recor = info_distancies_multi(mtrx_neta, llista_ciutats)
recor

#REVISAR OUTPUT perqùe NO s'estan càlculant bé les dades ja que aquestes no es poden ordenar bé perquè estem seleccionant múltiples
# amb el mateix index...
#NO ordena bé les ciutats perquè posa en l'index en les primeres posicions les ciutats que hem introdeuit en la llista


,Distància des de Córdoba
Distancia (kilómetros),
Barcelona,863.0
Valencia,523.0
Sevilla,143.0
Zaragoza,701.0
Málaga,159.0
Murcia,434.0
Bilbao,789.0
Alicante,507.0
Córdoba,NaN


***Funció de mapa amb folium*** -- idea de Francesco

*Desenvolupada per mi: Dioulde*

In [ ]:
# Para hacer un mapa de rutas, hauriem de seguir els següents passos:
#1) Installar la llibreria folium (pip install folium)
#2) Importar la llibreria
#3) Definir una funció que prengui com a entrada la matriu de distàncies i la llista de ciutats d'origen, i que retorni un mapa amb les rutes traçades.
    #3.1) Hauriem de importar les coordenades de les ciutats per poder traçar les rutes al mapa. Podem crear un diccionari amb les coordenades de cada ciutat. 
    #3.2) A la funció, primer crearíem un mapa base centrat en una ubicació específica (per exemple, la primera ciutat d'origen de la llista). 
    # Després, afegiríem marcadors per a cada ciutat d'origen i traçaríem línies entre les ciutats d'origen i les altres ciutats per mostrar les rutes.
    #3.3) Sí fos possible, podríem afegir informació addicional als marcadors, com per exemple la distància des de l'origen o la diferència de distància respecte a altres ciutats.
    #4) Finalment, retornaríem el mapa amb les rutes traçades.


<span style="font-size:20px; color:deepskyblue">**Referèncias:**</span><br>

* <span style="font-size:12px">[PANDAS - Manipulación de Datos con Python](https://joserzapata.github.io/courses/python-ciencia-datos/pandas/)

*  <span style="font-size:14px">[
DataFrames con Estilo: Formatos condicionales, colores y más.](https://www.youtube.com/watch?v=swzuFTMKAwk)</span>



<span style="font-size:16x; color:red">**POSAR TOTES LES REFERÈNCIES IT Academy i altres AQUÍ**